### ISSUE TITLE
Braille trace to update selected line message

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#4682]
**Reported by surfer0627 on 2014-12-11 02:47**
Hi, 
  Now, when user press shift+arrow down (few times) to select messages. Braille always update the first line.
expected: braille trace to update the selected line.
STR:
Here is a message cut in three lines.
Text message:
BlindSquare is the World’s Most Popular accessible GPS-app developed for the blind and visually 
impaired. It describes the environment, announces points of interest and street intersections as 
you travel. In conjunction with free, third-party navigation apps it is a powerful solution 

When pressing shift+arrow down (two times) to select lines, braille update the first line: "BlindSquare is the World’s Most Popular accessible GPS...".
expected: Braille updates the second line: "impaired. It describes the environment, announces points ..."
Thanks.

**Comment by dkager** [comment]
Another problem of the current implementation of braille.TextInfoRegion.update() is that it restricts itself to the line that contains the cursor. This prevents translating text into braille that the user can't scroll to without moving the cursor. We need to extend this so it moves to the *active* end of the selection, if any. Code to find the active end is part of #5770.

**Comment by dkager** [comment]
@jcsteh In braille.TextInfoRegion.update() I read:

```
		# HACK: Some TextInfos only support UNIT_LINE properly if they are based on POSITION_CARET,
		# so use the original cursor TextInfo for line and copy for cursor.
		self._readingInfo = readingInfo = self._getCursor()
		cursor = readingInfo.copy()
```

This is only used near the end of the function:

```
		# If this is not the start of the object, hide all previous regions.
		start = cursor.obj.makeTextInfo(textInfos.POSITION_FIRST)
		self.hidePreviousRegions = (start.compareEndPoints(readingInfo, "startToStart") < 0)
```

Why does this not use self.obj.makeTextInfo? It would simplify the code a bit if this could be avoided. If not it'd be great if the comment could be a bit clearer.

**Comment by jcsteh** [comment]
The comment you reference isn't at all related to the question you're 
asking. That doesn't mean we shouldn't add another comment though. :)

The reason is that self._getCursor might not return something created by 
self.obj directly. For example, ReviewTextInfoRegion._getCursor returns 
api.getReviewPosition(). We want to be sure we're always using the same 
object to make all TextInfos, even if _getCursor returned something from 
a different object for some reason.

**Comment by dkager** [comment]
> The comment you reference isn't at all related to the question you're asking.

Those are the very best questions!
Anyway, that local variable 'cursor' is only used in the bit of code I quoted. It seems to me that `readingInfo.obj == cursor.obj`, though readingInfo does get expanded along the way.

**Comment by jcsteh** [comment]
Oh. Now I see what you mean. So yes, readingInfo.obj and cursor.obj are 
equal, but you originally mentioned self.obj, which might not be equal.

Once upon a time, cursor was used instead of sel (before we supported 
selection for braille). Now that it isn't, there's no need for cursor 
once it's copied, but apparently, I never removed it. :)

**Comment by nishimotz** [comment]
According to deaf blind community in Japan, people who uses only braille display cannot check the ending character of selected text region, when the region contains several lines.

step to reproduce:

* put text to multiline edit (such as notepad.exe) which contains two or three lines of string
* NVDA+Ctrl+T to set "braille tether to focus"
* move to the beginning
* press shift + right arrow for several times. braille display shows the text and the selected region is underlined.
* press shift + right arrow and the caret goes to the second line. braille display only shows the text of first line.
* the press of shift + right arrow is accompanied by the speech announcement of newly selected character. however, the braille-only user cannot figure out where is the caret.
* during the procedure, scroll forward/back of braille display may be available, but sometimes such operations may fail and the selection is cancelled.

The workaround for this is firstly move to the ending of the region, and then, use shift + left arrow to increase the region in reverse. however, it is not intuitive.

Is this duplicate of #5770 or here?
Or is there any other solutions for this case and should be filed separately?

**Comment by jcsteh** [comment]
@dkager, am I correct that this is fixed by #6775? It seems to work for me, but I know you felt #5770 and this issue were a bit different.

**Comment by dkager** [comment]
I believe I initially wanted to do the selection code in two steps: first for left/right selections on a single line in #5770 and then multi-line selections in this issue. But in the end I did both, so yes this is fixed.
